envoy/router/router.h

```diff
@@ -530,12 +530,55 @@ using ShadowPolicyPtr = std::shared_ptr<ShadowPolicy>;
   STATNAME(vcluster)                                                                               \
   STATNAME(vhost)
 
+/**
+ * All route level stats. @see stats_macro.h
+ */
+#define ALL_ROUTE_STATS(COUNTER, GAUGE, HISTOGRAM, TEXT_READOUT, STATNAME)                         \
+  COUNTER(upstream_rq_retry)                                                                       \
+  COUNTER(upstream_rq_retry_limit_exceeded)                                                        \
+  COUNTER(upstream_rq_retry_overflow)                                                              \
+  COUNTER(upstream_rq_retry_success)                                                               \
+  COUNTER(upstream_rq_timeout)                                                                     \
+  COUNTER(upstream_rq_total)                                                                       \
+  STATNAME(route)                                                                                  \
+  STATNAME(vhost)
+
 /**
  * Struct definition for all virtual cluster stats. @see stats_macro.h
  */
 MAKE_STAT_NAMES_STRUCT(VirtualClusterStatNames, ALL_VIRTUAL_CLUSTER_STATS);
 MAKE_STATS_STRUCT(VirtualClusterStats, VirtualClusterStatNames, ALL_VIRTUAL_CLUSTER_STATS);
 
+/**
+ * Struct definition for all route level stats. @see stats_macro.h
+ */
+MAKE_STAT_NAMES_STRUCT(RouteStatNames, ALL_ROUTE_STATS);
+MAKE_STATS_STRUCT(RouteStats, RouteStatNames, ALL_ROUTE_STATS);
+
+/**
+ * RouteStatsContext defines config needed to generate all route level stats.
+ */
+struct RouteStatsContext {
+private:
+  const Stats::StatNameManagedStorage route_stat_name_storage_;
+  Stats::ScopeSharedPtr route_stats_scope_;
+
+public:
+  Stats::StatName route_stat_name_;
+  mutable RouteStats route_stats_;
+  RouteStatsContext(Stats::Scope& scope, const RouteStatNames& route_stat_names,
```

Comment:
nit: this is a complex constructor to inline into the header. There may not be an obvious cc file for this to go to, but you could try putting it into the main prod implementation file.

in any case, can you move the constructor above the two public member vars, and move the whole section to be above the private section?

Reply:
I moved the member variables. But I could n't think of moving the constructor to a better cc file